radix-rust/src/resolve.rs

```diff
@@ -0,0 +1,318 @@
+    fn labelled_resolve_from(value: &'a String, resolver: &impl LabelResolver<X>) -> X {
+        resolver.resolve_label_into(value.as_str())
+    }
+}
+
+impl<X: LabelledResolvable<ResolverOutput = X>> LabelledResolveFrom<String> for X {
+    fn labelled_resolve_from(value: String, resolver: &impl LabelResolver<X>) -> X {
+        resolver.resolve_label_into(value.as_str())
+    }
+}
+
+//==============================================================
+// Handle Option<X>
+//==============================================================
+// - None and Some(X) are handled by the identity above
+// - We then handle label -> Some(X) below
+//==============================================================
+
+impl<X: LabelledResolvable> LabelledResolvable for Option<X> {
+    type ResolverOutput = X;
+}
+
+impl<X: LabelledResolvable> LabelledResolveFrom<Option<X>> for Option<X> {
+    fn labelled_resolve_from(value: Option<X>, _resolver: &impl LabelResolver<X>) -> Option<X> {
+        value
+    }
+}
+
+impl<X: LabelledResolvable> LabelledResolveFrom<X> for Option<X> {
+    fn labelled_resolve_from(value: X, _resolver: &impl LabelResolver<X>) -> Option<X> {
+        Some(value)
+    }
+}
+
+impl<'a, X: LabelledResolvable + Clone> LabelledResolveFrom<&'a X> for Option<X> {
+    fn labelled_resolve_from(value: &'a X, _resolver: &impl LabelResolver<X>) -> Option<X> {
+        Some(value.clone())
+    }
+}
+
+impl<'a, X: LabelledResolvable + Clone> LabelledResolveFrom<&'a Option<X>> for Option<X> {
+    fn labelled_resolve_from(value: &'a Option<X>, _resolver: &impl LabelResolver<X>) -> Option<X> {
+        value.clone()
+    }
+}
+
+impl<'a, X: LabelledResolvable> LabelledResolveFrom<&'a str> for Option<X> {
+    fn labelled_resolve_from(value: &'a str, resolver: &impl LabelResolver<X>) -> Option<X> {
+        Some(resolver.resolve_label_into(value))
+    }
+}
+
+impl<'a, X: LabelledResolvable> LabelledResolveFrom<&'a String> for Option<X> {
+    fn labelled_resolve_from(value: &'a String, resolver: &impl LabelResolver<X>) -> Option<X> {
+        Some(resolver.resolve_label_into(value.as_str()))
+    }
+}
+
+impl<'a, X: LabelledResolvable> LabelledResolveFrom<String> for Option<X> {
+    fn labelled_resolve_from(value: String, resolver: &impl LabelResolver<X>) -> Option<X> {
+        Some(resolver.resolve_label_into(value.as_str()))
+    }
+}
+
+//==============================================================
+// Handle collections
+//==============================================================
+// - An iterator over something that resolves to X, resolves to
+//   the given collection/s of X.
+// Feel free to add more collections here as needed.
+//==============================================================
+
+impl<X: LabelledResolvable> LabelledResolvable for Vec<X> {
+    type ResolverOutput = X;
+}
+
+impl<'a, T: IntoIterator, X: LabelledResolvable<ResolverOutput = X>> LabelledResolveFrom<T>
```

Comment:
Minor: I understand that this is a matter of taste but:

1. To me these generic bounds are too complex to just write inline.
2. We're using a where clause already, we might as well just define the generic bounds there.

Reply:
I'm happy if you want to bring up this discussion as a team so we can revise our guidance about how to balance using bounds-at-definition, impl arguments and where clauses :)